### PR TITLE
feat(distinct): add distinct to query builder

### PR DESCRIPTION
**Pull request by OlegDO**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -212,22 +222,31 @@ class TypeormJsonQuery<TEntity = ObjectLiteral> {
...
-  public getAttributes(attrs: IJsonQuery<TEntity>['attributes'] = []): string[] {
-    const { isDisableAttributes } = this.options;
+  public getAttributes(attrs: IJsonQuery<TEntity>['attributes'] = []): IJsonQueryAttribute[] {
+    const { isDisableAttributes, distinctType } = this.options;
 
     const attributes = [
       ...(isDisableAttributes ? [] : this.query.attributes ?? []),
       ...attrs,
       ...(this.authQuery.attributes || []),
     ].map((field) => {
```
Comment (MatthewPattell) [inline src/index.ts]: you can use reduce e.g. and don't use lodash

File: package.json [hunk 2]
```diff
@@ -70,9 +70,11 @@
 		"ts-node": "^10.9.1",
 		"tsconfig-paths": "^4.2.0",
 		"ttypescript": "^1.5.15",
-		"typescript": "^4.7.4"
-	},
+		"typescript": "^4.7.4",
+    "@types/lodash": "^4.14.202"
+  },
 	"peerDependencies": {
+    "lodash": "^4.17.21",
```
Comment (MatthewPattell) [inline package.json]: don't use this

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 3]
```diff
@@ -212,22 +221,40 @@ class TypeormJsonQuery<TEntity = ObjectLiteral> {
...
+        name: this.withFieldAlias((typeof field === 'object' ? field.name : field) as string),
+        isDistinct:
+          distinctType === DistinctType.DISABLED || typeof field === 'string'
+            ? false
+            : Boolean(field?.isDistinct),
+      };
 
-    return [...new Set(attributes)];
+      // Get only unique attributes
+      const existingAttribute = iJsonQueryAttributes.find(
```
Comment (MatthewPattell) [inline src/index.ts]: we can do better:
```
const result = attr.reduce(res, { name, isDistinct }, () => {

  return {
   ...res,
   [field.name]: { name: ..., isDistinct  },
  }
});

return Object.values(result);
```

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 4]
```diff
@@ -212,22 +221,37 @@ class TypeormJsonQuery<TEntity = ObjectLiteral> {
...
-      }
+  public getAttributes(attrs: IJsonQuery<TEntity>['attributes'] = []): IJsonQueryAttribute[] {
+    const { isDisableAttributes, distinctType } = this.options;
+
+    return Object.values(
+      [
+        ...(isDisableAttributes ? [] : this.query.attributes || []),
+        ...attrs,
+        ...(this.authQuery.attributes || []),
+      ].reduce((res: object, field) => {
```
Comment (MatthewPattell) [inline src/index.ts]: We don't use object. Use Record or describe type

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 5]
```diff
@@ -212,22 +221,49 @@ class TypeormJsonQuery<TEntity = ObjectLiteral> {
...
-      ...attrs,
-      ...(this.authQuery.attributes || []),
-    ].map((field) => {
-      if (!field || typeof field !== 'string') {
-        throw new Error('Invalid json query: some attribute has incorrect name.');
-      }
-
-      return this.withFieldAlias(field);
-    });
+  public getAttributes(attrs: IJsonQuery<TEntity>['attributes'] = []): IJsonQueryAttribute[] {
```
Comment (MatthewPattell) [inline src/index.ts]: You can return: Rrquired<IJsonQueryAttribute>[]

File: src/index.ts [hunk 6]
```diff
@@ -1075,8 +1160,11 @@ class TypeormJsonQuery<TEntity = ObjectLiteral> {
     conditions.forEach((condition) => queryBuilder.andWhere(condition));
     groupByAttr.forEach((field) => queryBuilder.addGroupBy(field));
 
-    if (select.length) {
-      queryBuilder.select(select);
+    if (selectAttributes.length) {
+      // Build distinct only if enabled distinct and distinct attributes exists
+      distinctType !== DistinctType.DISABLED && isDistinctAttributesExist
+        ? this.applyDistinctSelectToQuery(queryBuilder, selectAttributes)
+        : queryBuilder.select(selectAttributes.map(({ name }) => name as string));
```
Comment (MatthewPattell) [inline src/index.ts]: Then here remove: as string

File: src/index.ts [hunk 7]
```diff
@@ -212,22 +221,49 @@ class TypeormJsonQuery<TEntity = ObjectLiteral> {
...
+
+    return Object.values(
+      [
+        ...(isDisableAttributes ? [] : this.query.attributes || []),
+        ...attrs,
+        ...(this.authQuery.attributes || []),
+      ].reduce(
+        (
+          res: Record<
+            IJsonQueryAttribute['name'],
```
Comment (MatthewPattell) [inline src/index.ts]: You can just use string

File: src/index.ts [hunk 8]
```diff
@@ -212,22 +221,49 @@ class TypeormJsonQuery<TEntity = ObjectLiteral> {
...
+            { name: IJsonQueryAttribute['name']; isDistinct: boolean }
```
Comment (MatthewPattell) [inline src/index.ts]: Here also

File: src/index.ts [hunk 9]
```diff
@@ -1075,8 +1160,11 @@ class TypeormJsonQuery<TEntity = ObjectLiteral> {
     conditions.forEach((condition) => queryBuilder.andWhere(condition));
     groupByAttr.forEach((field) => queryBuilder.addGroupBy(field));
 
-    if (select.length) {
-      queryBuilder.select(select);
+    if (selectAttributes.length) {
+      // Build distinct only if enabled distinct and distinct attributes exists
+      distinctType !== DistinctType.DISABLED && isDistinctAttributesExist
```
Comment (MatthewPattell) [inline src/index.ts]: Use if block

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 5 above)
Comment (OlegDO) [reply src/index.ts]: Here is another problem with type. I will add IAttribute internal interfaces, and if we will modify IJonQueryAttribute we will not have redundant fields in result of get attributes
image

File: src/index.ts (hunk 9 above)
Comment (OlegDO) [reply src/index.ts]: I'll create separate method for avoid nested if block

**Review by MatthewPattell — APPROVED** [review]

**Pull request merged**